ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java

```diff
@@ -0,0 +1,411 @@
+  private Runnable queueCallback = () -> { };
+
+  public PullQueryWriteStream(
+      final OptionalInt hardLimit,
+      final StreamedRowTranslator translator
+  ) {
+    this.hardLimit = hardLimit;
+    this.translator = translator;
+
+    // register a drainHandler that will wake up anyone waiting on hasCapacity
+    drainHandler.add(ignored -> {
+      monitor.enter();
+      monitor.leave();
+    });
+  }
+
+  private static final class HandledRow {
+    private final PullQueryRow row;
+    private final Handler<AsyncResult<Void>> handler;
+
+    private HandledRow(
+        final PullQueryRow row,
+        final Handler<AsyncResult<Void>> handler
+    ) {
+      this.row = row;
+      this.handler = handler;
+    }
+  }
+
+  // -------------------- READ QUEUE METHODS -------------------------
+
+  @Override
+  public void drainTo(final Collection<? super KeyValueMetadata<List<?>, GenericRow>> collection) {
+    monitor.enter();
+    try {
+      while (!queue.isEmpty()) {
+        collection.add(poll());
+      }
+    } finally {
+      monitor.leave();
+    }
+  }
+
+  public void drainRowsTo(final Collection<PullQueryRow> collection) {
+    monitor.enter();
+    try {
+      while (!queue.isEmpty()) {
+        collection.add(pollRow());
+      }
+    } finally {
+      monitor.leave();
+    }
+  }
+
+  @Override
+  public KeyValueMetadata<List<?>, GenericRow> poll(
+      final long timeout,
+      final TimeUnit unit
+  ) throws InterruptedException {
+    if (monitor.enterWhen(hasData, timeout, unit)) {
+      try {
+        return poll();
+      } finally {
+        monitor.leave();
+      }
+    }
+
+    return null;
+  }
+
+  @Override
+  public KeyValueMetadata<List<?>, GenericRow> poll() {
+    final PullQueryRow row = pollRow();
+    if (row == null) {
+      return null;
+    }
+
+    if (row.getConsistencyOffsetVector().isPresent()) {
+      return new KeyValueMetadata<>(RowMetadata.of(row.getConsistencyOffsetVector().get()));
+    }
+
+    return new KeyValueMetadata<>(
+        KeyValue.keyValue(null, row.getGenericRow()),
+        row.getSourceNode()
+            .map(sn -> new KsqlHostInfo(sn.getHost(), sn.getPort()))
+            .map(RowMetadata::of)
+    );
+  }
+
+  public PullQueryRow pollRow(
+      final long timeout,
+      final TimeUnit unit
+  ) throws InterruptedException {
+    if (monitor.enterWhen(hasData, timeout, unit)) {
+      try {
+        return pollRow();
+      } finally {
+        monitor.leave();
+      }
+    }
+
+    return null;
+  }
+
+  private PullQueryRow pollRow() {
+    final HandledRow polled;
+    monitor.enter();
+    try {
+      polled = queue.poll();
+    } finally {
+      monitor.leave();
+    }
+
+    if (polled == null) {
+      return null;
+    }
+
+    polled.handler.handle(FUTURE_FACTORY.succeededFuture());
+
+    monitor.enter();
+    try {
+      if (queue.size() < softLimit) {
+        drainHandler.forEach(h -> h.handle(null));
```

Comment:
Also, are you meant to empty the drain handler once entries are called back or do they need only be registered once?  Looking through our codebase, I'm not sure -- some things register every time the writer is at capacity, and others register just once.

Reply:
that's a question I had as well, I'm not entirely sure which should be the case - any preference?

Reply:
I've been using `NetSocketImpl` as my inspiration and they do not remove the drainhandler after it was called, so I'll keep it that way